### PR TITLE
[RF] Simplify HistFactory models

### PR DESCRIPTION
**Pull request by hageboeck**
Here, HistFactory is streamlined a bit.

The models are restructured from this:
[HFTest4_old.pdf](https://github.com/root-project/root/files/6466419/HFTest4_old.pdf)

To this:
[HFTest4.pdf](https://github.com/root-project/root/files/6466420/HFTest4.pdf)

The main difference is that first, all binned parts of the calculation are executed (in batches), and then all scale factors are applied to the fully interpolated binned distributions. This reduces the number of function calls and allows for non-uniform binning (ROOT-4958), since the bin width correction is not applied as the last step, but during the binned calculations.

Further, prototypes of batch evaluations functions were added for all objects in HistFactory calculations. There are some possibilities for optimisations:
- [ ] ParamHistFunc::evaluateSpan always recalculates all indices
- [ ] RooBinIntegrator doesn't support multi-dimensional batch evaluations
- [ ] Auto-vectorisation should be checked, especially for performance-critical things like the piecewise interpolation
- [ ] Profiling with an example like in `testHistfactory` but with more bins should be done.

### CODE REVIEW

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @hageboeck, thanks a lot for these very nice developments! I went through it commit-by-commit to understand what the code is doing, ran the tests an the tutorials, and everything seems good to me.

I still made a few comments mostly about technical details, it would be nice if you could address these before we merge next week. We also have some other large PRs in the pipeline (multiprocessing from Patrick and batchmode from Manos), and I would like to merge this PR before.

Cheers!

File: roofit/roofitcore/src/RooHistFunc.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooHistFunc.cxx:239]: That's nice. I would actually suggest to put an assert into that function that checks that the preconditions are met. I costs minimal run time in debug builds, but it will help a lot in debugging if you violate the preconditions.
I'll add such a commit.

Comment (hageboeck) [reply roofit/roofitcore/src/RooHistFunc.cxx:239]: I take that back, because it's already happening.

Comment (hageboeck) [reply roofit/roofitcore/src/RooHistFunc.cxx:239]: Actually no. It was only checking the size. Commit coming.

File: roofit/roofitcore/src/RooAbsData.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsData.cxx:1775]: I didn't think about touching it because I was in minimally invasive mode. I'll give it a try.

File: roofit/roofitcore/src/RooBinIntegrator.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:30]: `IsA()` comes in through the `ClassDef(..., 1)` macro. It's not in TClass. What it the exact error that you see?

Comment (hageboeck) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:30]: Ah, you cannot call TClass::GetName() because it's not included?

File: roofit/roofitcore/src/RooBinIntegrator.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:137]: Sure, I didn't consider this in the scope of this PR, but I'll do you the favour.

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (hageboeck) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx:495]: Note that this isn't c++11. I can do it for you, but it requires including ROOT's manual backport.

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (hageboeck) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx:415]: And if that wasn't clear: Not my code! ;-)

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (hageboeck) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx:761]: - Not c++11! **Don't do this in interfaces!!!** Not even when ROOT becomes c++14 ...
- Not sure if vectors automatically convert to spans?
- One could do this with `begin(), end()`, but I wouldn't add another constructor. You can do if you want, but I'll go now with `begin, end`.

Comment (guitargeek) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx:761]: Right, it's not C++11. But could we not use the backport of `std::span` in ROOT, like you did for the RooSpan back in the day?
https://github.com/root-project/root/blob/master/roofit/batchcompute/inc/RooSpan.h#L20
But okay, maybe we should better not do interface changes to basic RooFit classes in this PR. If they aren't completely non-controversial. I'm taking my suggestion back.

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (guitargeek) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx:495]: Yes, the manual backport is already widely used in RooFit

File: roofit/roofitcore/src/RooBinIntegrator.cxx
Comment (guitargeek) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:137]: Thanks!

File: roofit/roofitcore/src/RooBinIntegrator.cxx
Comment (guitargeek) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:30]: Sorry, I was not precise here. correct, I mean the call to `TClass::GetName()` right after `IsA`:
```
/home/rembserj/root-builds/histFactory/src/roofit/roofitcore/src/RooBinIntegrator.cxx: In static member function ‘static void RooBinIntegrator::registerIntegrator(RooNumIntFactory&)’:
/home/rembserj/root-builds/histFactory/src/roofit/roofitcore/src/RooBinIntegrator.cxx:59:68: error: invalid use of incomplete type ‘class TClass’
   59 |   RooNumIntConfig::defaultConfig().method1D().setLabel(proto->IsA()->GetName()) ;
```

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (hageboeck) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx]: No, not in interfaces. Note that it's always `RooSpan` in all interfaces. Imagine this situation:
- We call something `std::xxx` in ROOT
- People include it, and it works, because that thing is not part of the `std::` (because it's c++11).
- People switch their project to `c++20`, because "Why not?". ROOT's interfaces are c++11, but that will be compatible.
- Now there's two definitions of `std::span` in the interfaces. Nothing will compile.

You simply *cannot* use those things in interfaces. If it's internal to ROOT, it's OK, because we can typedef or fix includes etc, but once something is in the interfaces, it's out in the world.
Our span thing works because once ROOT goes c++20, you just exchange the (non-standard) `std::span` inside RooSpan with the real thing, and fix a few lines of code in RooSpan if necessary. Interfaces remain unchanged, and everything will just work.

**Comment by hageboeck** [comment]
@guitargeek I think I addressed everything, but let me know about that TClass include. When those are OK, I'll do a proper rebase and we let the CI run one final time.

File: roofit/roofitcore/src/RooBinIntegrator.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooBinIntegrator.cxx:30]: OK, that's indeed a violation of "include what you use". I'll add it back.

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (hageboeck) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx]: I fixed the template `(begin, end)` constructor of `RooArgSet`, so I was now able to write almost the exact same line as you proposed.

File: roofit/roofitcore/src/RooAbsData.cxx
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsData.cxx:37]: I think you need to keep this: `strlcpy` is used twice in this file, and without the include it doesn't compile for me

File: roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx
Comment (guitargeek) [reply roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx]: That's a nice solution! And thanks for explaining the situation with the `std::` backports in interfaces.

**Comment by guitargeek** [comment]
Hi Stephan, thanks for the updates! I went over all the new commits, so from my side it's okay now to squash and rebase.

There are just two includes that are missing for it to compile on my system at this point:
* `#include "strlcpy.h"` in roofit/roofitcore/src/RooAbsData.cxx
* `#include "TClass.h"` in roofit/roofitcore/src/RooBinIntegrator.cxx

**Review by guitargeek — APPROVED** [review]
Since my last comment, the missing includes were added, the commits were squashed and the code was rebased. I already reviewed the actual contents of the commits before.

So, all good to me, thanks a lot!

File: roofit/roofitcore/src/RooAbsData.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsData.cxx:37]: I fixed it by going to RooHelpers::tokenise to tokenise the strings.

**Comment by hageboeck** [comment]
@guitargeek I (tried to) fix a clang-tidy warning about a possible `nullptr` by changing a loop in `RooAbsData.cxx`. Let's wait for this run and hopefully merge after that. hopefully.

**Comment by guitargeek** [comment]
Okay, thanks!

**Comment by guitargeek** [comment]
Thanks for the updates again! The jenkins tests passed, so we can merge it now. Note that there are some remaining Travis CI failures, but as far as I can tell they are not related to the changes in this PR. They relate to the `checkLimits()` call in `RooBinIntegrator`, which has already been there before.

**Pull request merged**
